### ISSUE TITLE
App updates may not come from the intended remote

### ISSUE THREAD

**Issue by mwleeds** [flatpak/flatpak#1447]
You should not be able to update an app such that its collection ID changes, but you can.
```
$ flatpak info com.endlessm.wiki_art.en
Ref: app/com.endlessm.wiki_art.en/x86_64/eos3
ID: com.endlessm.wiki_art.en
Arch: x86_64
Branch: eos3
Origin: eos-apps
Collection ID: com.endlessm.Dev.Apps
Date: 2018-01-26 21:32:56 +0000
Subject: Export com.endlessm.wiki_art.en
Commit: 160436e78fce0a5f4576c5baa06b0492971faebccd91839d1b67dc06ece106de
Parent: fd3d1be2bcc34d2200fbc1b47857714a47270fd741c3bc677a10e0771bd04198
Location: /var/lib/flatpak/app/com.endlessm.wiki_art.en/x86_64/eos3/160436e78fce0a5f4576c5baa06b0492971faebccd91839d1b67dc06ece106de
Installed size: 13.1 MB
Runtime: com.endlessm.apps.Platform/x86_64/3
```
```
$ cat /ostree/repo/config
[core]
repo_version=1
mode=bare
add-remotes-config-dir=false
min-free-space-percent=0

[remote "eos"]
url=https://ostree.endlessm.com/ostree/eos-amd64
branches=os/eos/amd64/eos3;
xa.disable=true

[remote "flathub"]
url=https://dl.flathub.org/repo/
xa.title=Flathub
gpg-verify=true
gpg-verify-summary=true

[remote "eos-apps"]
url=https://ostree.endlessm.com/ostree/eos-apps
xa.default-branch=eos3
gpg-verify=true
gpg-verify-summary=false
xa.title=Endless Applications
collection-id=com.endlessm.Apps

[remote "eos-runtimes"]
url=https://ostree.endlessm.com/ostree/eos-amd64
gpg-verify=true
gpg-verify-summary=false
xa.title=Endless OS and Runtimes
collection-id=com.endlessm.Os

[remote "eos-sdk"]
url=https://ostree.endlessm.com/ostree/eos-sdk
gpg-verify=true
gpg-verify-summary=false
xa.title=Endless Runtime and SDK
collection-id=com.endlessm.Sdk
```
Running `G_MESSAGES_DEBUG=all flatpak update com.endlessm.wiki_art.en`: 
[flatpak-update-diff-collection-try2.log](https://github.com/flatpak/flatpak/files/1749830/flatpak-update-diff-collection-try2.log)
```
$ flatpak info com.endlessm.wiki_art.en
Ref: app/com.endlessm.wiki_art.en/x86_64/eos3
ID: com.endlessm.wiki_art.en
Arch: x86_64
Branch: eos3
Origin: eos-apps
Collection ID: com.endlessm.Apps
Date: 2018-01-26 21:32:56 +0000
Subject: Export com.endlessm.wiki_art.en
Commit: 59b85f128df8a8841257f67c38b314d4a8450a8853036c53d32427795db3cb52
Parent: d7f0c3937383db6688266bebd42adc4a2a0e76ef1d26f4aa03118a16f3ee2bdc
Location: /var/lib/flatpak/app/com.endlessm.wiki_art.en/x86_64/eos3/59b85f128df8a8841257f67c38b314d4a8450a8853036c53d32427795db3cb52
Installed size: 13.1 MB
Runtime: com.endlessm.apps.Platform/x86_64/3
```

**Comment by mwleeds** [comment]
Steps to reproduce: (sorry these aren't really possible for people outside Endless)
1. Flash a computer with `eos-eos3.3-amd64-amd64.180221-070058.base.iso`
2. Run `flatpak remote-modify --collection-id=com.endlessm.Apps eos-apps; flatpak remote-modify --collection-id=com.endlessm.Sdk eos-sdk; flatpak remote-modify --collection-id=com.endlessm.Os eos-runtimes`
3. Update WikiArt, either with the flatpak CLI or GNOME Software

**Comment by pwithnall** [comment]
To be clear, I think this is an issue because an attacker could trick you into adding a remote for some shiny new app, but that remote then also serves updates to other apps which you didn’t expect.

I guess I would be happier about the collection ID changing if `flatpak` errored out if the old and new versions of the app you’re updating are signed with different GPG keys. What currently happens in that situation?

**Comment by mwleeds** [comment]
> To be clear, I think this is an issue because an attacker could trick you into adding a remote for some shiny new app, but that remote then also serves updates to other apps which you didn’t expect.
> 
> I guess I would be happier about the collection ID changing if flatpak errored out if the old and new versions of the app you’re updating are signed with different GPG keys. What currently happens in that situation?

I think the short answer to your question is that attack is only possible when `--user` is used. When the system helper is used for deployment, `handle_deploy` calls `flatpak_dir_pull_untrusted_local` which calls `ostree_repo_verify_commit_for_remote` using the remote of the ref being updated, so that keyring is used.

When `--user` was used, flatpak relies on ostree's GPG verification during the pull. In that case (and for users of ostree other than flatpak) the attack is only possible if the sketchy remote is configured with the same collection ID as the legitimate one. `OstreeRepoFinderConfig` [only checks remotes with the same collection ID as the ref](https://github.com/ostreedev/ostree/blob/1c9baad8a886399ef2362420fe34c4718d06f0de/src/libostree/ostree-repo-finder-config.c#L145) and that remote's keyring is then used by `find_remotes_cb` (and in the pull code if it makes it that far).

For the similar attack vector of advertising malicious refs over LAN or USB, the answer is similar. Since `OstreeRepoFinderMount` and `OstreeRepoFinderAvahi` both use `ostree_repo_resolve_keyring_for_collection()`, the only way a ref could be verified with the "wrong" keyring is if two remotes have the same collection ID, and in that case the behavior depends on the order of the remotes.

So the status quo isn't great. Some options are:
1. Make it an error to have more than one remote with the same collection ID (which I guess means you can't configure mirrors, but ostree already has metalink support).
2. Make it an error to update a ref in a way that changes the keyring used to sign it (which seems like it should be done at the level of ostree not flatpak, and implies there can never be a mechanism for key rotation).
3. Accept the current behavior, and have front-ends like GNOME Software warn the user if a new remote is duplicating an existing collection ID (but you'd also have to consider the case of the collection ID being deployed after the remote is added).

cc: @alexlarsson @cgwalters

**Comment by pwithnall** [comment]
Thanks for investigating.

> So the status quo isn't great.

From the sounds of it, the status quo is pretty good. As I understand it (correct me if I’m wrong), with `--user` the sketchy remote has to have the same collection ID as the existing remote for that app in order to be able to provide upgrades for an app. Without `--user`, the sketchy remote can have a different collection ID, but must have the same GPG key as the existing remote.

To double-check, is it the case that an app can be upgraded to a version from a repository with a different collection ID if (and only if) the other repository uses the same GPG keys as the old one?

> Some options are:
> 
>  1. Make it an error to have more than one remote with the same collection ID (which I guess means you can't configure mirrors, but ostree already has metalink support).

A non-starter. Configuring two remotes to have the same collection ID is part of a valid migration path for old repository configurations to new ones (the other part of that approach being to merge the repositories on the server and redirect their URIs to the same place).

> 2. Make it an error to update a ref in a way that changes the keyring used to sign it (which seems like it should be done at the level of ostree not flatpak, and implies there can never be a mechanism for key rotation).

I suspect that updating a ref from a commit signed with one key to a commit signed with a different key should be an error, yes, unless explicitly allowed by the user or by the old key (for example, signing a new set of GPG keys with the old key and adding them to commit metadata).

I think flatpak has a mechanism for GPG key rotation already by adding a `xa.gpg-keys` key to the `ostree-metadata`.

> 3. Accept the current behavior, and have front-ends like GNOME Software warn the user if a new remote is duplicating an existing collection ID (but you'd also have to consider the case of the collection ID being deployed after the remote is added).

We possibly should do that.

**Comment by ramcq** [comment]
I believe on the way to deciding on collection IDs, we discussed and rejected the idea that GPG keys should be used as a way to identify commits, on the basis that they are non distinct and can validly change over the lifetime of a repo.

A deployed Flatpak currently stores the name of the remote it came from. A locally configured remote binds together some information, such as the URL we pull from by default, a collection ID we will permit for alternative sources (essentially: match these possible mirrors) and the GPG key which must be present when we pull from that remote.

We shouldn't be looking elsewhere. I don't think we need to look at the last commit - and nor do we want to. If we've automatically changed to a new key in the repo configuration, or the user has changed it (omg old key compromised and actively expoited by hackers) then it would be incorrect to allow a signed commit with the old key, even if we have an app installed with that old key.

As long as we always enforce the locally-configured GPG key for remote, this isn't an attack vector, I don't think. I think if --user pull allows us to pick up (even if they are validly signed for /that/ remote) commits from a different remote with the same collection ID, that's a bug which should be fixed.

(So I don't believe we need to warn about clashing collection IDs in the eg GNOME Software UI for human beings, but we might want to print a warning on the command line.)

**Comment by mwleeds** [comment]
Despite this issue stagnating for a while I think there is a real problem and I have an idea for a solution. The problem is that the configured remotes aren't all trusted equally, but the P2P code (`ostree_repo_find_remotes_async`/`ostree_repo_pull_from_remotes_async`) doesn't respect this. For example if I install an app from flathub I expect the updates for it to be signed with the GPG key on the flathub remote, regardless of if there's another remote that has its collection ID set to `org.flathub.Stable` (perhaps a remote controlled by a malicious actor). This set of constraints runs counter to the way `OstreeRepoFinderConfig` works.

So the solution I'm proposing is to add an option that can be passed to `ostree_repo_find_remotes_async`, something like `origin-remote` or `trusted-remote` that could be set to `flathub` for example. Then `OstreeRepoFinderConfig` would only return that remote, and the pull code would only use the GPG keys on that remote. As I said above, I believe this will only effectively be a change in behavior for `--user` pulls.

I haven't written the code yet, but that makes sense right?

**Comment by matthiasclasen** [comment]
Why do you need to call ostree_repo_find_remotes_async when you already know that it will return flathub ?

**Comment by mwleeds** [comment]
Because it may also return USB drives or LAN peers

**Comment by matthiasclasen** [comment]
wait, then .. you don't know what it'll return after all. Maybe your description of the 'trusted remote' concept was just not clear enough. This trusted remote is used for
* GPG keys
* collection ID

Anything else ?

**Comment by mwleeds** [comment]
I discussed this with @pwithnall over email and here's the alternative fix being considered:

Make it an error to configure a remote to have the same collection ID as an existing remote (either at the time the remote is added by the user or at the time the configuration is being updated based on metadata keys) **unless** the GPG keys of the two remotes also match. It would also have to be an error to update the GPG keys associated with a remote in such a way that the remotes would have matching collection IDs and differing GPG keys (in case for example the old keys are compromised, a malicious actor could switch to the compromised keys, deploy the collection ID `org.flathub.Stable`, and then switch back to their own keys).

I think the above fix needs to be implemented because otherwise once flatpak/ostree switch to using `refs/mirrors` more than `refs/remotes`, you would end up with issues if `refs/mirrors/org.flathub.Stable/ostree-metadata` could be signed with GPG Key A or GPG Key B depending on which remote it was pulled from.

But even with that it might make sense to implement the `trusted-remote` support for the purpose of layered security, and in case users of libostree other than flatpak have the same model of not all remotes being equally trustworthy.

**Comment by mwleeds** [comment]
> > But even with that it might make sense to implement the `trusted-remote` support for the purpose of layered security, and in case users of libostree other than flatpak have the same model of not all remotes being equally trustworthy.
> 
> I am concerned about this, and disagree with the idea of `trusted-remotes`. As I understand it, it limits the scope of P2P pulls from ever being parallelised more than non-P2P pulls.

I don't see the connection to parallelization. A `trusted-remotes` option on `ostree_repo_find_remotes_async` would only dictate what GPG keyring each result should use (precisely, a result that provides refs[0] should use the keyring associated with trusted_remotes[0])

>The problem here is one of preventing malicious configuration from ever being stored in the user’s repo config, not one of trying to prevent that malicious configuration from causing damage after it’s already been added to the repo config.

Preventing the config change from happening (at least from flatpak's perspective) is the goal of #2601 but as I said in that commit message it's not a perfect solution. And I'm not convinced it makes sense to leave the pull code defenseless in case for example a new remote is dropped into `/etc/flatpak/remotes.d/` (in which case I think Flatpak has no opportunity to check the correctness of the new remote).

> I’m not really clear on how we got from [suggestion 3](https://github.com/flatpak/flatpak/issues/1447#issuecomment-371450010) to the much-more-radical “Then I tried adding an interface, `OstreeRepoFinder2` which replaces `ostree_repo_finder_resolve_async`”. It would be good to have a summary of the chain of thinking so far (including a summary of our e-mail conversation, so it’s public and all discussion is summarised in one place). Currently, I come back to this task every other week and find it hard to work out where we were before, what’s changed, and what’s informing your current thinking.

I agree this thread is hard to follow; sorry about that. Here's my attempt at summarizing the chain of thought for all of this:

- This issue started out being about the supposed problem that an app can change from one collection ID to another when being updated. However that alone is not an issue; an app may legitimately change from staging to prod or vice versa, and those changes may have different collection IDs.
- The potential problem is if the P2P code allows an app to be updated from a remote other than the one it was installed from. It turns out this is possible if the sketchy remote has the same collection ID configured as the legitimate remote.
- Disallowing more than remote from having the same collection ID is not a good solution, because there are some legitimate reasons that could happen (e.g. when two repositories are being merged into one)
- However there should be no legitimate reason for two remotes to have the same collection ID and differing GPG keys; refs from the same collection should be signed with the same keys.
- If we want to disallow remotes to clash in that way, #2601 is a start but libostree should probably have a similar mechanism
- Since there are ways to get around the above check, it might make sense for the pull code to be robust against such a scenario of a less trusted remote having the same collection ID as the legitimate one. This is a problem at the level of libostree, and could be solved in `ostree_repo_find_remotes_async` but that would mean any user of `ostree_repo_finder_resolve_async` and the related API would be "on their own" and expected to decide for themselves which results are legitimate.
- To change `ostree_repo_finder_resolve_async` would mean changing the `OstreeRepoFinder` interface (in other words deprecating and replacing it) but doing that is a massive change, so perhaps there's an easier way.

While writing that I had another idea: this could be solved by adding an option to `ostree_repo_pull_from_remotes_async` because you have to use that API to pull regardless of whether you used `ostree_repo_find_remotes_async` or `ostree_repo_finder_resolve_async` or `ostree_repo_finder_resolve_all_async` to get the results.

**Comment by mwleeds** [comment]
Here's an excerpt from the email conversation which seems like the interesting part:

Philip: 

> My point is that if someone’s compromised your local config, you have bigger problems. It’s fine for two remotes to share the same collection ID (as long as they offer the same content); it’s not fine for a random repo to be able to get a collection ID into your local config which matches one already there, without the user explicitly acking that.


me:

> 1. We're not really talking about a whole config being compromised; we're talking about one remote being compromised. And in Flatpak, [each remote is trusted to a different level](https://github.com/flatpak/flatpak/pull/2091#issuecomment-427357671). So a compromise of one shouldn't affect the apps installed from the others
> 
> 2. I don't think it's workable to require user interaction if a remote tries to deploy a collection ID to the same value as an existing one because (a) the update could be happening in the background without any UI and (b) it's probably unrealistic to expect users to know what a collection ID is.
> 
> On the other hand you could use the heuristic "A remote cannot set a collection ID the same as an existing one if its GPG keys differ" but then you also have to do a similar check any time the GPG keys are updated. I wonder if there's any use case that would be broken by that. Perhaps the migration path you mentioned above?

(But I don't think a merge of two repos would be broken by #2601. By the time a collection ID is deployed, they should both be using the same GPG keys.)
